### DISCUSSION TITLE
Running full stack tests against a connected DB.

### DISCUSSION THREAD

**Discussion by william-scalexp** [blitz-js/blitz#2433] [Ideas]
### What do you want and why?
BlitzJS looks amazing and I love the work done in bridging the frontend/backend.

What would make BlitzJS even more awesome is the ability to  connect frontend tests against 
a test database instead of using mocks. 

Why would this be amazing?
- Test the frontend against a realistic backend.
- Test the exposed backend queries with realistic calls from the frontend.
- Faster than using cypress (and less flakey) for end to end tests.
- Releverage test data factories/scenarios.

### Possible implementation(s)

Not sure!

**Comment by flybayer** [comment]
Hey! New apps are generated with a setup for that. `.env.test.local` has a unique test database url. So when you run `NODE_ENV=test blitz start` for example it will use the test DB.

I use Cypress with this way in a client app for running e2e tests. I'm wanting to look at switching to Playwright though.

So I think most of your points are already possible/addressed. Except for your Cypress point, what do you mean by that?

**Reply by uptickmetachu** [reply]
I was hoping we could test using react-testing-library instead of cypress. Aka using JSDom tests without backend mocks this might require modifying the api shimming layer at test time.

**Reply by flybayer** [reply]
Ok, I think you should be able to do that. If you run into a specific problem, then definitely open an issue for it.

**Reply by william-scalexp** [reply]
From a fresh skeleton app;

I've modified app/pages/index.test.tsx to look like this (removed the mocks).

```
import { render } from "test/utils"

import db from "db"
import Home from "./index"
import { useCurrentUser } from "app/core/hooks/useCurrentUser"


describe("renders blitz documentation link", () => {
  // This is an example of how to ensure a specific item is in the document
  // But it's disabled by default (by test.skip) so the test doesn't fail
  // when you remove the the default content from the page

  // This is an example on how to mock api hooks when testing
  it('does something', async() => {
    const { getByText } = render(
      <Home />
    )

    const linkElement = getByText(/Documentation/i)

    expect(linkElement).toBeInTheDocument()

  })
})
``` 

This component uses a `Query` to get the current user. 

The error I get is:
```
  ● renders blitz documentation link › does something

    TypeError: Cannot read property 'apiUrl' of undefined

      3 |
      4 | export const useCurrentUser = () => {
    > 5 |   const [user] = useQuery(getCurrentUser, null)
        |                  ^
      6 |   return user
      7 | }
      8 |
  ```
  
  I'm assuming the query components aren't configured to be run this way in a test. Thoughts on this sort of workflow? 
  This would be more akin to being able to test `Views` as in Django or other monolithic frameworks. 
  
  The testing workflow would involve: 
  1. Starting a transaction (to roll back any database changes)
  1. Create a test scenario using Database Model factories
  1. Interact with it in an end to end fashion (without cypress)
  1. Roll back the transaction.
  
  In theory this would make integration testing much quicker and easier. This however would only work if all api calls are made within the same transaction. This isn't possible with cypress because we are operating at the browser level.

**Reply by flybayer** [reply]
Ok got it. I think what we could do in jest tests is bypass our RPC code. So `getCurrentUser` would be invoked directly as a function. That's the main issue here from what I can tell. Can you open an issue specifically for this?

And yeah it would be good to add some type of factories and things.

With prisma I don't think we can do it all in transactions like that. What I'm currently doing is resetting the DB on each test like this:

```ts
beforeEach(async () => {
  await db.$reset()
})
```

**Reply by maxfi** [reply]
> I'm wanting to look at switching to Playwright though.

@flybayer Have you switched to playwright already? I'm interested in setting up playwright to test a blitz app and was wondering if there are any best practices for integration with jest and the blitz jest preset?

**Reply by flybayer** [reply]
No I haven't :)

**Reply by maxfi** [reply]
@flybayer You should check it out when you have a moment. With the new inbuilt recording functionality it's relatively easy to get up and running quickly. 🚀  I've managed to get a mixed integration/e2e test up and running in my blitz app. I'm using playwright in jest and am simply [requiring playwright](https://playwright.dev/docs/test-runners#jest--jasmine) in my jest tests. This is working well but ideally I'd like to seperate the e2e tests from my unit tests to avoid running them all the time. I'm not sure how to configure this with the `blitz` jest preset. If I move the e2e tests into `test/e2e` then I can avoid running them with the unit tests but then I have to use [`@playwright/test`](https://playwright.dev/docs/1.12.0/test-intro) with `npx playwright test` and also can't import my blitz code any longer - which is useful for setting up the database before running playwright tests. Would it be worth it to start another discussion about this? TIA. 🙏

**Reply by flybayer** [reply]
@maxfi yeah let's take that to a new discussion. But probably you want to tweak the jest projects config to have a new e2e project: https://github.com/blitz-js/blitz/blob/canary/packages/blitz/jest-preset.js#L41

**Reply by maxfi** [reply]
Thanks for that info @flybayer. Let me have a play around with that and get back to you in a new discussion.
